### PR TITLE
[SYCL][NFC] Handle a bunch of warnings coming from SYCL RT

### PR DESCRIPTION
**Pull request by alexbatashev**
- `-Wstring-conversion`: get rid of converting `const char*` to `bool` in `assert`
- `-Winconsistent-missing-override`: add missing `final` or `override`
- `-Wcovered-switch-default`: remove unreachable `default` clauses
- `-Wpessimizing-move`: allow temporary object copy elision by removing `std::move`
- `-Wcast-qual`: explicitly cast away `const` qualifier
- `-Wunused-parameter`: remove some unused parameters

### CODE REVIEW

**Comment by bader** [comment]
> * `-Wstring-conversion`: get rid of converting `const char*` to `bool` in `assert`
> * `-Winconsistent-missing-override`: add missing `final` or `override`
> * `-Wcovered-switch-default`: remove unreachable `default` clauses
> * `-Wpessimizing-move`: allow temporary object copy elision by removing `std::move`
> * `-Wcast-qual`: explicitly cast away `const` qualifier
> * `-Wunused-parameter`: remove some unused parameters

CI system must prevent us from adding new warnings enabled by these options.

**Comment by alexbatashev** [comment]
> > * `-Wstring-conversion`: get rid of converting `const char*` to `bool` in `assert`
> > * `-Winconsistent-missing-override`: add missing `final` or `override`
> > * `-Wcovered-switch-default`: remove unreachable `default` clauses
> > * `-Wpessimizing-move`: allow temporary object copy elision by removing `std::move`
> > * `-Wcast-qual`: explicitly cast away `const` qualifier
> > * `-Wunused-parameter`: remove some unused parameters
> 
> CI system must prevent us from adding new warnings enabled by these options.

I'm afraid, some of those are specific to newer compiler version than what we have in CI.

File: sycl/source/detail/scheduler/commands.hpp [hunk 1]
```diff
@@ -422,8 +422,8 @@ class MemCpyCommand : public Command {
 
   void printDot(std::ostream &Stream) const final;
   const Requirement *getRequirement() const final { return &MDstReq; }
-  void emitInstrumentationData();
-  ContextImplPtr getContext() const override final;
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/commands.hpp:426]: Why is `override` removed?
AFAIK `final` and `override` are independent.

`getContext()` is `virtual`.

Comment (bader) [reply sycl/source/detail/scheduler/commands.hpp:426]: FYI: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-override

File: sycl/include/CL/sycl/detail/image_accessor_util.hpp [hunk 2]
```diff
@@ -945,19 +931,19 @@ DataT ReadPixelDataLinearFiltMode(const cl_int8 CoordValues,
 
   // Get Color Values at each Coordinate.
   cl_float4 Ci0j0k0 = getColorInFloat(cl_int4{i0, j0, k0, 0});
-
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/detail/image_accessor_util.hpp:948]: Do you really need these stylistic changes in in this pull-request?

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/image_accessor_util.hpp:948]: I think, no. My editor does it automatically. I'll revert these.

File: sycl/include/CL/sycl/detail/image_accessor_util.hpp [hunk 3]
```diff
@@ -196,9 +196,6 @@ vec<T, 4> readPixel(T *Ptr, const image_channel_order ChannelOrder,
     Pixel.y() = Ptr[2]; // g
     Pixel.x() = Ptr[3]; // r
     break;
-  default:
-    throw cl::sycl::invalid_parameter_error("Unhandled image channel order",
-                                            PI_INVALID_VALUE);
```
Comment (garimagu) [inline sycl/include/CL/sycl/detail/image_accessor_util.hpp:201]: Why was this removed? Can we add something else? Will an assert also give a warning?

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/image_accessor_util.hpp:201]: @garimagu that code handles all cases from an enum. It'd never reach `default:` branch anyway.

Comment (garimagu) [reply sycl/include/CL/sycl/detail/image_accessor_util.hpp:201]: Ok.. So you get this warning because of this "-Wcovered-switch-default: remove unreachable default clauses" warning?

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/image_accessor_util.hpp:201]: Right, this is a new warning, introduced in recent GCC versions, that's why it's not hitting us in CI.

File: sycl/include/CL/sycl/detail/image_accessor_util.hpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/image_accessor_util.hpp]: Done.

File: sycl/include/CL/sycl/detail/image_accessor_util.hpp (hunk 3 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/image_accessor_util.hpp:201]: @garimagu please, unresolve this conversations if you still have questions.
